osu.Game/Overlays/Profile/Sections/Recent/DrawableRecentActivity.cs

```diff
@@ -0,0 +1,147 @@
+using osu.Game.Online.API;
+using osu.Game.Online.API.Requests;
+using osu.Game.Online.Chat;
+using osu.Game.Screens.Select.Leaderboards;
+
+namespace osu.Game.Overlays.Profile.Sections.Recent
+{
+    public class DrawableRecentActivity : DrawableProfileRow
+    {
+        private APIAccess api;
+
+        private readonly RecentActivity activity;
+        private readonly string userLinkTemplate;
+        private readonly string beatmapLinkTemplate;
+        private readonly string beatmapsetLinkTemplate;
+
+        private LinkFlowContainer content;
+
+        public DrawableRecentActivity(RecentActivity activity)
+        {
+            this.activity = activity;
+
+            userLinkTemplate = $"[{urlToAbsolute(activity.User?.Url)} {activity.User?.Username}]";
+            beatmapLinkTemplate = $"[{urlToAbsolute(activity.Beatmap?.Url)} {activity.Beatmap?.Title}]";
+            beatmapsetLinkTemplate = $"[{urlToAbsolute(activity.Beatmapset?.Url)} {activity.Beatmapset?.Title}]";
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(APIAccess api)
+        {
+            this.api = api;
+
+            LeftFlowContainer.Add(content = new LinkFlowContainer
+            {
+                AutoSizeAxes = Axes.Y,
+                RelativeSizeAxes = Axes.X,
+            });
+
+            RightFlowContainer.Add(new OsuSpriteText
+            {
+                Text = activity.CreatedAt.LocalDateTime.ToShortDateString(),
+                Anchor = Anchor.TopRight,
+                Origin = Anchor.TopRight,
+                Font = "Exo2.0-RegularItalic",
+                TextSize = 12,
+                Colour = OsuColour.Gray(0xAA),
+            });
+
+            var formatted = MessageFormatter.FormatText(activityToString());
+
+            content.AddLinks(formatted.Text, formatted.Links);
+        }
+
+        protected override Drawable CreateLeftVisual()
+        {
+            switch (activity.Type)
+            {
+                case RecentActivityType.Rank:
+                    return new DrawableRank(activity.ScoreRank)
+                    {
+                        RelativeSizeAxes = Axes.Y,
+                        Width = 60,
+                        FillMode = FillMode.Fit,
+                    };
+
+                case RecentActivityType.Achievement:
+                    return new MedalIcon(activity.Achievement.Slug)
+                    {
+                        RelativeSizeAxes = Axes.Y,
+                        Width = 60,
+                        FillMode = FillMode.Fit,
+                    };
+
+                default:
+                    return new Container
+                    {
+                        RelativeSizeAxes = Axes.Y,
+                        Width = 60,
+                        FillMode = FillMode.Fit,
+                    };
+            }
+        }
+
+        private string urlToAbsolute(string url) => $"{api?.Endpoint ?? @"https://osu.ppy.sh"}{url}";

```

Comment:
this seems unnecessary, but i'm unsure whether taking endpoint from api or hardcoding is preferred. or whether to use `osu://` links.

Reply:
using the api is fine, but there's no case the api is null, so the fallback is not required.